### PR TITLE
Remove obtain_from_generic and use .as<T> instead

### LINKED ISSUE #4574: Remove `obtain_from_generic` from all classes
We now have `.as<T>` which does the job. Easy entrance task

### PR DESCRIPTION
**Pull request by remyabel**
Fixes #4574

### CODE REVIEW

File: examples/meta/src/statistical_testing/linear_time_maximum_mean_discrepancy.sg [hunk 1]
```diff
@@ -49,7 +49,7 @@ real width = learnt_kernel_single.get_real("log_width")
 #![select_kernel_combined]
 mmd.set_kernel_selection_strategy(enum EKernelSelectionMethod.KSM_MAXIMIZE_POWER, True)
 mmd.select_kernel()
-CombinedKernel learnt_kernel_combined = CombinedKernel:obtain_from_generic(mmd.get_kernel())
+CombinedKernel learnt_kernel_combined = mmd.get_kernel()
```
Comment (karlnapf) [inline examples/meta/src/statistical_testing/linear_time_maximum_mean_discrepancy.sg]: .get_real_vector(“xxx”)
Where you need to check for xxx for the kernel weights are registered in CCombinedKernel.cpp (search for SG_ADD)

File: src/shogun/distributions/Gaussian.cpp [hunk 2]
```diff
@@ -439,17 +439,3 @@ SGVector<float64_t> CGaussian::sample()
...
-{
-	if (!distribution)
-		return NULL;
-
-	CGaussian* casted=dynamic_cast<CGaussian*>(distribution);
-	if (!casted)
-		return NULL;
-
-	/* since an additional reference is returned */
-	SG_REF(casted);
```
Comment (karlnapf) [inline src/shogun/distributions/Gaussian.cpp]: careful here! if this is used, there will be a memory problem if you use as<T> as it doesnt increase the ref counter

File: src/shogun/evaluation/CrossValidation.h [hunk 3]
```diff
@@ -57,25 +57,6 @@ namespace shogun
...
-		{
-			if (!eval_result)
-				return NULL;
-
-			REQUIRE(
-			    eval_result->get_result_type() == CROSSVALIDATION_RESULT,
-			    "CrossValidationResult::obtain_from_generic(): argument is"
-			    "of wrong type!\n");
-
-			SG_REF(eval_result);
```
Comment (karlnapf) [inline src/shogun/evaluation/CrossValidation.h]: same here, need to be careful with memory errors

File: src/shogun/evaluation/GradientResult.h [hunk 4]
```diff
@@ -43,23 +43,6 @@ class CGradientResult : public CEvaluationResult
...
-	 * @return casted CGradientResult object
-	 */
-	static CGradientResult* obtain_from_generic(CEvaluationResult* eval_result)
-	{
-		ASSERT(eval_result);
-
-		REQUIRE(eval_result->get_result_type()==GRADIENTEVALUATION_RESULT,
-				"Provided evaluation result is not of type CGradientResult!\n")
-
-		SG_REF(eval_result);
```
Comment (karlnapf) [inline src/shogun/evaluation/GradientResult.h]: and here

File: src/shogun/modelselection/GradientModelSelection.cpp [hunk 5]
```diff
@@ -173,8 +173,7 @@ float64_t CGradientModelSelection::get_cost(SGVector<float64_t> model_vars, SGVe
 
 	// evaluate the machine
 	CEvaluationResult* evaluation_result=m_machine_eval->evaluate();
-	CGradientResult* gradient_result=CGradientResult::obtain_from_generic(
-			evaluation_result);
+	CGradientResult* gradient_result = evaluation_result->as<CGradientResult>();
```
Comment (karlnapf) [inline src/shogun/modelselection/GradientModelSelection.cpp]: here we go: the previous function did a `SG_REF`, but now it is missing -> problems

File: examples/meta/src/statistical_testing/linear_time_maximum_mean_discrepancy.sg [hunk 6]
```diff
@@ -49,7 +49,7 @@ real width = learnt_kernel_single.get_real("log_width")
...
 RealVector weights = learnt_kernel_combined.get_subkernel_weights()
```
[the remark this answers is not in the document]
Comment (remyabel) [reply examples/meta/src/statistical_testing/linear_time_maximum_mean_discrepancy.sg]: Thanks.

**Comment by remyabel** [comment]
I believe I've managed to address your concerns. PTAL.

**Comment by karlnapf** [comment]
closed?

**Pull request closed without merging**